internal/server/repository_workerauth.go

```diff
@@ -159,14 +159,39 @@ func StoreNodeInformationTx(ctx context.Context, reader db.Reader, writer db.Wri
 			WithReleaseVersion(fmt.Sprintf("Boundary v%s", wci.BoundaryVersion)),
 		)
 		wConf.Type = PkiWorkerType.String()
-		wConf.PublicId, err = newWorkerIdFromScopeAndName(ctx, scopeId, wci.Name)
+		wConf.PublicId, err = NewWorkerIdFromScopeAndName(ctx, scopeId, wci.Name)
 		if err != nil || wConf.PublicId == "" {
 			return errors.Wrap(ctx, err, op, errors.WithMsg("error creating a worker id"))
 		}
 		workerCreateConflict := &db.OnConflict{
 			Target: db.Columns{"public_id"},
 			Action: db.SetColumns([]string{"release_version", "name", "description"}),
 		}
+
+		// Before going ahead with this, we need to see if this is an upgrade
+		// case from the old KMS auth method. If so we need to delete first as
+		// type is immutable.
+		lookupWorker := allocWorker()
+		lookupWorker.PublicId = wConf.PublicId
+		err := reader.LookupById(ctx, &lookupWorker)
+		switch {
+		case err == nil:
+			if lookupWorker.Type == KmsWorkerType.String() {
+				// Delete it
+				rowsDeleted, err := writer.Delete(ctx, lookupWorker)
```

Comment:
Not really...if a worker somehow changes its type _while it's already running_ then... 🤷 .

Generally this would happen because they shut the worker down, upgrade it, and on next start it does this.